### PR TITLE
Situational plays beta

### PR DESCRIPTION
**Pull request by liquidmetal9015**
## Description
Some simple changes to fix the preempt function and get the scores and preemption actually working. This will need some testing and there might be an cleaner way to do the score function.

NOTE: Situation analysis is enabled by default in this currently, we might want it to be disabled by default for merging into staging/master

## Associated Issue
Issue #1342

## Design Documents
[Link](link-to-design-doc)

### Test Case 1
1. Run soccer with both adaptive formation and basic_122 enabled

Expected result: The plays switch dynamically according to the situation lists in each play

### CODE REVIEW

**Comment by liquidmetal9015** [comment]
Make sure to run this and check for the correct behavior.

File: soccer/gameplay/plays/offense/adaptive_formation.py
Comment (tjones320) [inline soccer/gameplay/plays/offense/adaptive_formation.py:161]: What is this doing now???

File: soccer/gameplay/plays/offense/basic_122.py
Comment (tjones320) [inline soccer/gameplay/plays/offense/basic_122.py]: So can super score see this?

File: soccer/gameplay/situational_play_selection.py
Comment (tjones320) [inline soccer/gameplay/situational_play_selection.py]: add more comments

File: soccer/gameplay/situational_play_selection.py
Comment (tjones320) [inline soccer/gameplay/situational_play_selection.py:355]: add more comments

File: soccer/gameplay/standard_play.py
Comment (tjones320) [inline soccer/gameplay/standard_play.py:80]: I really like this.

**Review by tjones320 — APPROVED** [review]

File: soccer/gameplay/plays/offense/adaptive_formation.py
Comment (JNeiger) [inline soccer/gameplay/plays/offense/adaptive_formation.py]: I personally would prefer these to be enums so I don't have to worry about misspelling them

Comment (tjones320) [reply soccer/gameplay/plays/offense/adaptive_formation.py]: I think that's a fair point

Comment (tjones320) [reply soccer/gameplay/plays/offense/adaptive_formation.py]: I agree

Comment (kylestach) [reply soccer/gameplay/plays/offense/adaptive_formation.py]: Thirded.

File: soccer/gameplay/plays/offense/basic_122.py
Comment (kylestach) [reply soccer/gameplay/plays/offense/basic_122.py]: I would prefer an `is_valid(situation)` function instead of just a public list.

File: soccer/gameplay/situational_play_selection.py
Comment (kylestach) [inline soccer/gameplay/situational_play_selection.py]: Is this the expected branch? Shouldn't the situation always be an enum value or am I missing something.

File: soccer/gameplay/standard_play.py
Comment (kylestach) [reply soccer/gameplay/standard_play.py:80]: What about it, exactly?

Comment (kylestach) [reply soccer/gameplay/standard_play.py:80]: Also, we need to have a lot more documentation about what the overridden class method needs to do.

File: soccer/gameplay/situational_play_selection.py
Comment (kylestach) [inline soccer/gameplay/situational_play_selection.py:344]: I would prefer this have some interaction with the play itself. i.e. `if currentPlay.tryPreempt(self.currentSituation): # Try preempt`. But I'm open to changing my mind on this.

**Review by kylestach — COMMENTED** [review]
Mostly looks pretty good. Needs more documentation considering this is now an important part of the plays system's API.

File: soccer/gameplay/plays/offense/adaptive_formation.py
Comment (liquidmetal9015) [reply soccer/gameplay/plays/offense/adaptive_formation.py]: So currently both

`situationList = [ 'clear', 'defend_goal', 'defensive_scramble', 'offensive_pileup','defensive_pileup']`

and
```
situationList = [situational_play_selection.SituationalPlaySelector.Situation.CLEAR,
            situational_play_selection.SituationalPlaySelector.Situation.DEFEND_GOAL,
            situational_play_selection.SituationalPlaySelector.Situation.DEFENSIVE_SCRAMBLE,
            situational_play_selection.SituationalPlaySelector.Situation.OFFENSIVE_PILEUP,
            situational_play_selection.SituationalPlaySelector.Situation.DEFENSIVE_PILEUP ]
```

are tested and working. I can remove string support if we want; I mostly added it because I thought the enums would be long although I suppose that's fine. It does have an option to spell check the strings as a default parameter. Having to write everything out brings two things to my mind though: one should I make the enum its own file?, and also I am open to renaming some things. situational_play_selection and SituationalPlaySelector might be both too wordy and no longer accurate to what the file and class do. Situation Analysis has been what we have been calling it but I'm open to suggestions.

File: soccer/gameplay/plays/offense/basic_122.py
Comment (liquidmetal9015) [reply soccer/gameplay/plays/offense/basic_122.py]: I can make an is_valid but would still prefer making an "private" _situationList to pass to situation analysis and to make it so that all you have to do is override that variable to make everything work at a basic level and allow you to use the super score function as is. is_valid would basically just be a contains call though if I'm understanding right.

File: soccer/gameplay/situational_play_selection.py
Comment (liquidmetal9015) [reply soccer/gameplay/situational_play_selection.py:344]: Currently planning on adding
```
    def try_preempt(self):
        self.terminate()
        return True
```
to standard play that can be overridden if you want more complex functionality, which will also remove preempt logic from execute_running and change update_preempt to be actually calling for preempts rather than just updating a value.

File: soccer/gameplay/plays/offense/adaptive_formation.py
Comment (liquidmetal9015) [reply soccer/gameplay/plays/offense/adaptive_formation.py:161]: This calls the score function in standard play, if it returns none then you are supposed to use your own score function, if not it gives you the starting point for your score based on the situation list you have made for the class. I'll add more documentation before merge.

File: soccer/gameplay/situational_play_selection.py
Comment (kylestach) [reply soccer/gameplay/situational_play_selection.py:344]: That sounds good to me.

File: soccer/gameplay/plays/offense/basic_122.py
Comment (kylestach) [reply soccer/gameplay/plays/offense/basic_122.py]: That sounds like the correct fix.

File: soccer/gameplay/plays/offense/adaptive_formation.py
Comment (liquidmetal9015) [inline soccer/gameplay/plays/offense/adaptive_formation.py]: So uhh, this is how make pretty wants to do it right now. See my comment on moving the enum out of situational_play_selection? Or import in a different way?

File: soccer/gameplay/situational_play_selection.py
Comment (liquidmetal9015) [inline soccer/gameplay/situational_play_selection.py:329]: I deprecated string support as it was confusing and adding complexity.

**Comment by liquidmetal9015** [comment]
Ok, @kylestach @JNeiger @tjones320. This should be ready for another pass of review.

File: soccer/gameplay/plays/offense/adaptive_formation.py
Comment (kylestach) [reply soccer/gameplay/plays/offense/adaptive_formation.py]: Interesting formatting. Does it do this when you manually put each entry on a separate line?

File: soccer/gameplay/plays/offense/adaptive_formation.py
Comment (kylestach) [reply soccer/gameplay/plays/offense/adaptive_formation.py:161]: Honestly, I would rather have this hook into the system in a manner other than the `score` function - if you do it this way you have to properly implement the super-call.

I would _prefer_ the play selection explicitly check against `is_applicable` for each play, and only check `score` if `is_applicable` returns true.

File: soccer/gameplay/situational_play_selection.py
Comment (kylestach) [reply soccer/gameplay/situational_play_selection.py:329]: :+1:, agree with this change.

File: soccer/gameplay/situational_play_selection.py
Comment (kylestach) [inline soccer/gameplay/situational_play_selection.py]: This could just be `return self.currentSituation in situations`, right?

File: soccer/gameplay/situational_play_selection.py
Comment (kylestach) [inline soccer/gameplay/situational_play_selection.py]: Shouldn't this use the result from `try_preempt()`? Or does it just use the fact that the play has now been terminated if preemption succeeded?

File: soccer/gameplay/standard_play.py
Comment (kylestach) [inline soccer/gameplay/standard_play.py]: Addition to

File: soccer/gameplay/situational_play_selection.py
Comment (kylestach) [inline soccer/gameplay/situational_play_selection.py:327]: Documentation

**Review by kylestach — CHANGES_REQUESTED** [review]

File: soccer/gameplay/plays/offense/adaptive_formation.py
Comment (liquidmetal9015) [reply soccer/gameplay/plays/offense/adaptive_formation.py]: Yes. They started on individual lines, I just tested it again.

File: soccer/gameplay/situational_play_selection.py
Comment (liquidmetal9015) [reply soccer/gameplay/situational_play_selection.py]: Current behavior is just that it will continuously call try_preempt() until the play changes without regard to the return value of try_preempt(). Which I think is fine for the time being.

Comment (kylestach) [reply soccer/gameplay/situational_play_selection.py]: Okay. So the play is responsible for deciding if it's terminated, and then situational play selection chooses a new one if it terminates itself.

Sounds good to me.

File: soccer/gameplay/plays/offense/adaptive_formation.py
Comment (kylestach) [reply soccer/gameplay/plays/offense/adaptive_formation.py]:
```python
_situationList = [
    situational_play_selection.SituationalPlaySelector.Situation.ATTACK_GOAL,
    situational_play_selection.SituationalPlaySelector.Situation.DEFEND_CLEAR,
    situational_play_selection.SituationalPlaySelector.Situation.OFFENSIVE_SCRAMBLE
] # yapf: disable
```
Should do what you want.

**Review by kylestach — APPROVED** [review]
LGTM

**Comment by liquidmetal9015** [comment]
Hypothetically someone else should review this since a lot has changed. To that end I will update the how to test methodology.

## How to test
1. Set the enabled Boolean to true in situational_play_selection.py 
2. Enable both basic_122 and adaptive_formation at the same time
3. Move the ball from from and too each side of the field and changing possession

Result: The running play should change based on the valid situations for that play, you should see the scores changing in real time and plays should properly be preempted a short time after the situation changes.

(Random thought: The delay after situation change that it takes before preempting should probably have some mechanism of increasing for repeated transitions. Like if we go from a->b then b->a it should increase the delay and a->b increase it again to help reduce any waffling, might need to use some history to prevent cycles of 3 or more. That would be a future feature)

@tjones320 @JNeiger should both probably take a look at this again.

There will still be future iterations where some things will change. Modifying the score functionality in root_play to categorically exclude plays is something that is on the table. For the moment it has been decided to leave it up to the score function. Sticking for basics right now until we have a decent little pool of plays to be swapping between

File: soccer/gameplay/plays/offense/adaptive_formation.py
Comment (JNeiger) [reply soccer/gameplay/plays/offense/adaptive_formation.py:161]: Just in general something to keep in mind, the less boilerplate that needs to be added to new plays the better. It's much easier to mess with and see what a play is doing if you don't have to skip over tons of repeated stuff. If something is really simple, it should look really simple functionality wise.

File: soccer/gameplay/situational_play_selection.py
Comment (JNeiger) [inline soccer/gameplay/situational_play_selection.py]: `fudoablse`? Typo

File: soccer/gameplay/situational_play_selection.py
Comment (JNeiger) [inline soccer/gameplay/situational_play_selection.py:344]: Correct me if I'm wrong, but this will preempt on every situation change even if the current play accounts for this situation? Is that the behavior we want? I personally would let the current play continue to run so we don't loose state info from just restarting it.

File: soccer/gameplay/situational_play_selection.py
Comment (JNeiger) [inline soccer/gameplay/situational_play_selection.py:375]: This only asks the play if it should try to preempt once on situation change. Should we continuously do it? I imagine we might want to put off a preempt if a pass is happening or something, but will it be up to the play to then quit after that happens or should we just continue polling until that pass is finished

File: soccer/gameplay/standard_play.py
Comment (JNeiger) [inline soccer/gameplay/standard_play.py:82]: If you want to simplify the score for the time being, it might be easier to just return `0` here, and then in the plays, always return score + whatever was there before. This way you don't have that super ugly if statement in the score function of plays. If you want to keep it separated due to clarity about what's being executed when situationAnalysis is off or on thats very reasonable too

**Review by JNeiger — APPROVED** [review]
Functionality looks good, just some clarification on minut details

File: soccer/gameplay/situational_play_selection.py
Comment (liquidmetal9015) [reply soccer/gameplay/situational_play_selection.py:375]: So I had it preempting continuously up until a few days ago and I changed it to only do it a single time to as a quick fix to concerns I had about not detecting when a play is preempted and restarts, and continuously preempting it due to that (because we only check root play equality, which will still be true after switching). I plan to switch back to continuously preempting in the next iteration alongside a more robust detection for changes. This is just a beta that is disabled by default.

File: soccer/gameplay/situational_play_selection.py
Comment (liquidmetal9015) [reply soccer/gameplay/situational_play_selection.py:344]: So I think that on one hand if we are running in a gimped situation where we have one play covering many adjacent situations and are using our current plays, we will get a performance benefit from not preempting switches from and to the same play. But I think at this point it's more valuable to preempt them and see how everything handles it since that will be the expectation when we have good play coverage, our plays should be able to handle being started and stopped because that is how it's going to be with more plays in the playbook. For that reason I have kept it preempting switches between the same plays.

File: soccer/gameplay/standard_play.py
Comment (liquidmetal9015) [reply soccer/gameplay/standard_play.py:82]: I'm still thinking about how to make the score functions cleaner, probably by making some changes in how play selection works and changing the standard_play score function. Right now I think this is the safest way to go since if it is an no-run by default.

**Pull request merged**
